File: apps/web/src/components/results/jsonFormat.ts
export function formatJsonLabel(key: string): string {
  return key.replace(/_/g, ' ').replace(/-/g, ' ');
}

export function formatJsonValue(value: unknown, maxLength = 240): string {
  if (value === null || value === undefined) return '—';
  if (typeof value === 'boolean') return value ? 'yes' : 'no';
  if (typeof value === 'number' || typeof value === 'string') {
    const text = String(value);
    return text.length <= maxLength ? text : `${text.slice(0, maxLength)}…`;
  }
  if (Array.isArray(value)) {
    if (value.length === 0) return '[]';
    if (value.every((item) => ['string', 'number', 'boolean'].includes(typeof item))) {
      return value.map((item) => formatJsonValue(item, 80)).join(', ');
    }
    return `[${value.length} items]`;
  }
  if (typeof value === 'object') {
    return `{${Object.keys(value as object).length} fields}`;
  }
  return String(value);
}

export function unionTableColumns(rows: Array<Record<string, unknown>>): string[] {
  const columns: string[] = [];
  for (const row of rows) {
    for (const key of Object.keys(row)) {
      if (!columns.includes(key)) columns.push(key);
    }
  }
  return columns;
}

export function isScalar(value: unknown): boolean {
  return (
    value === null ||
    value === undefined ||
    typeof value === 'string' ||
    typeof value === 'number' ||
    typeof value === 'boolean'
  );
}

export function isMapList(value: unknown): value is Array<Record<string, unknown>> {
  return (
    Array.isArray(value) &&
    value.length > 0 &&
    value.every((item) => item !== null && typeof item === 'object' && !Array.isArray(item))
  );
}

File: apps/mobile/lib/widgets/results/json_format.dart
import 'dart:convert';

String formatJsonLabel(String key) {
  return key.replaceAll('_', ' ').replaceAll('-', ' ');
}

String formatJsonValue(dynamic value, {int maxLength = 240}) {
  if (value == null) return '—';
  if (value is bool) return value ? 'yes' : 'no';
  if (value is num || value is String) {
    final text = value.toString();
    if (text.length <= maxLength) return text;
    return '${text.substring(0, maxLength)}…';
  }
  if (value is List) {
    if (value.isEmpty) return '[]';
    if (value.every((item) => item is num || item is String || item is bool)) {
      return value.map((item) => formatJsonValue(item, maxLength: 80)).join(', ');
    }
    return '[${value.length} items]';
  }
  if (value is Map) {
    return '{${value.length} fields}';
  }
  return value.toString();
}

String formatJsonValueDetailed(dynamic value) {
  if (value is Map || value is List) {
    return const JsonEncoder.withIndent('  ').convert(value);
  }
  return formatJsonValue(value);
}

List<String> unionTableColumns(Iterable<Map> rows) {
  final columns = <String>[];
  for (final row in rows) {
    for (final key in row.keys) {
      final name = '$key';
      if (!columns.contains(name)) columns.add(name);
    }
  }
  return columns;
}

bool isScalar(dynamic value) {
  return value == null ||
      value is String ||
      value is num ||
      value is bool;
}

bool isMapList(dynamic value) {
  return value is List &&
      value.isNotEmpty &&
      value.every((item) => item is Map);
}